packages/bibliography/src/main.rs

```diff
@@ -0,0 +1,358 @@
+                        "imports": {"type": "set", "access": "add"}
+                    }
+                }
+            ]
+            }
+        )
+    );
+}
+
+// Cyclic dependencies.
+// Consider this document
+// [cite] a
+// [cite] b
+// If cite is both unknown_content and push access to the list, it means a must happen before
+// b (due to the order they appear in the document, list access is granular).
+// a would expand to one [list-push] and something else, but then we have a [list-push] which is
+// above b, which also requires list pushing to the same list. the [list-push] from a must thus
+// occur before b, but also, b is unknown-content and so it must happen before a. Thus, we have
+// a cyclic dependency.
+
+fn transform(from: &str, to: &str) {
+    let input: Value = {
+        let mut buffer = String::new();
+        io::stdin().read_to_string(&mut buffer).unwrap();
+        serde_json::from_str(&buffer).unwrap()
+    };
+
+    match from {
+        "cite" => transform_cite(to, &input),
+        "cite-internal-do-not-use" => transform_cite_label(to, &input),
+        "bibliography" => transform_bibliography(to, &input),
+        other => {
+            eprintln!("Package does not support {other}");
+        }
+    }
+}
+
+// The name Citation clashes with Hayagriva
+#[derive(Debug, Clone, Serialize, Deserialize)]
+struct InlineCitation<'a> {
+    key: &'a str,
+    note: Option<&'a str>,
+}
+
+fn transform_cite(_to: &str, input: &Value) {
+    let citation = {
+        let key = input["data"].as_str().unwrap();
+        let note = {
+            let arg = input["arguments"]["note"].as_str().unwrap();
+            (!arg.is_empty()).then_some(arg)
+        };
+        let citation = InlineCitation { key, note };
+        serde_json::to_string(&citation).unwrap()
+    };
+
+    let output = json!([push_citation!(&citation), cite_label!(&citation)]);
+
+    println!("{output}");
+}
+
+fn transform_cite_label(_to: &str, input: &Value) {
+    let citation = input["data"].as_str().unwrap();
+
+    let labels: Vec<String> = {
+        let var = env::var("inline_citation_labels").unwrap_or("[]".to_string());
+        serde_json::from_str(&var).unwrap()
+    };
+
+    if let Some(label) = labels.iter().find_map(|s| s.strip_prefix(citation)) {
+        println!("{label}");
+    } else {
+        eprintln!("You must add a [bibliography] for [cite]s to work. If you don't want it visible, do [bibliography visibility=hidden]");
+    }
+}
+
+fn transform_bibliography(_to: &str, input: &Value) {
+    let Some(bibliography) = read_bibliography(input) else { return; };
+
+    // This is the citations (i.e [cite]s) used in the text
+    let citations: Vec<String> = {
+        let var = env::var("inline_citations").unwrap_or("[]".to_string());
+        serde_json::from_str(&var).unwrap()
+    };
+
+    // This is our database of all entries (like a db containing all biblatex-entries)
```

Comment:
This can contain yaml entries too right?
```suggestion
    // This is our database of all entries
```

Reply:
Yes, it may indeed contain yaml entries. This was more of an explanation, "this is our database, behaving like a database of all biblatex-entries" to understand what it was a database of and get a better grasp on what it is doing. But I'll remove that part of the comment, sure!